### DISCUSSION TITLE
any'ness of id's - how to use it in a generic client

### DISCUSSION THREAD

**Discussion by lathoub** [FraunhoferIOSB/FROST-Server#659] [Q&A] [answered]
Hi,

I'm making a generic JavaScript [client for the OGC SensorThings API](https://github.com/lathoub/UISTA) and stumbled on the _any'ness_ of id (as per [spec](http://docs.opengeospatial.org/is/15-078r6/15-078r6.html#23)).

The USGS service has alphanumeric id's and the GET for Things is https://labs.waterdata.usgs.gov/sta/v1.1/Things('USGS-09213700')
The BGS service uses numeric id's and the GET looks like this: https://sensors.bgs.ac.uk/FROST-Server/v1.1/Things(10).

A quoted call to https://sensors.bgs.ac.uk/FROST-Server/v1.1/Things('10') results in a 404 `Not a valid id` (similar for an unquoted call to the USGS service).

Question: What is a good way to deal with this? Especially for a generic client that does not know if id are numeric or alphanumeric.
A numeric test (`isNaN`) in the client? (What is a service has both numeric and alphanumeric id's in the same table?)

Thanks!
Lathoub

**Comment by hylkevds** [comment, marked as answer]
You can't have different IDs in the same table / entity type, they're either a Number or a String.

FROST Client (java) does indeed test if the JSON entity in the parse tree is a number or not:

https://github.com/FraunhoferIOSB/FROST-Client/blob/cc8508e1108743412261a5bf069f8593624a5747/src/main/java/de/fraunhofer/iosb/ilt/sta/jackson/IdDeserializer.java#L35-L39
```
        JsonNode node = parser.getCodec().readTree(parser);
        if (node.isNumber()) {
            return new IdLong(node.asLong());
        }
        return new IdString(node.asText());
```

The different implementations of the {ID interface](https://github.com/FraunhoferIOSB/FROST-Client/blob/cc8508e1108743412261a5bf069f8593624a5747/src/main/java/de/fraunhofer/iosb/ilt/sta/model/Id.java) know how to output these IDs in URLs and JSON, though the `getUrl()` and `getJson()` methods.

**Reply by lathoub** [reply]
thanks @hylkevds . Solved!

https://github.com/lathoub/UISTA/blob/1458fbeaf03dc8188a31faf544bcb21b22277cad/resources/static/js/map.js#L35-L37

**Comment by hylkevds** [comment]
Cool that you're making a generic JavaScript client!
The in-development version 2 of FROST-Server can have different ID types per-table, so you could have Things with UUIDs and Observations with Long IDs.

**Reply by lathoub** [reply]
My client will be read-only, focusing on browsing and comparing multiple datastreams
